src/main/scala/org/constellation/domain/healthcheck/HealthCheckConsensus.scala

```diff
@@ -0,0 +1,569 @@
+  implicit val peerDataOrder: Order[PeerData] = Order.fromOrdering
+
+  val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+  val fetchProposalAfter = 120L
+  val removeUnresponsivePeersWithParallelRoundAfter = 60L
+
+  private[healthcheck] val roundIds: Ref[F, NonEmptySet[RoundId]] = Ref.unsafe(NonEmptySet.one(roundId))
+  private[healthcheck] val roundPeers: Ref[F, Map[Id, RoundData]] = {
+    Ref.unsafe(
+      initialRoundPeers
+        .mapValues(pd => RoundData(pd.asRight[ProxyPeers]))
+    )
+  }
+  private[healthcheck] val leftDuringConsensus: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val removedUnresponsivePeersWithParallelRound: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val ownConsensusHealthStatus: Ref[F, ConsensusHealthStatus] = Ref.unsafe(
+    ConsensusHealthStatus(
+      checkedId,
+      ownId,
+      roundId,
+      healthCheckStatus,
+      initialRoundPeers.mapValues(_.peerMetadata.nodeState)
+    )
+  )
+  private[healthcheck] val parallelRounds: Ref[F, Map[Id, Set[RoundId]]] = Ref.unsafe(Map.empty)
+
+  private val sendingProposalLock = createSemaphore()
+  private val fetchingProposalLock = createSemaphore()
+
+  def getTimeInSeconds(): F[Long] = C.monotonic(SECONDS)
+
+  def start(): F[Fiber[F, Unit]] = F.start(sendProposal())
+
+  def getOwnPerceivedHealthStatus(): F[ConsensusHealthStatus] = ownConsensusHealthStatus.get
+
+  def getRemovedUnresponsivePeersWithParallelRound(): F[Set[Id]] = removedUnresponsivePeersWithParallelRound.get
+
+  def addRoundIds(ids: Set[RoundId]): F[Unit] =
+    roundIds.modify(rounds => (NonEmptySet(rounds.head, rounds.tail ++ ids), ()))
+
+  def getRoundIds(): F[NonEmptySet[RoundId]] = roundIds.get
+
+  def getParallelRounds(): F[Map[Id, Set[RoundId]]] = parallelRounds.get
+
+  def addParallelRounds(id: Id, roundIds: Set[RoundId]): F[Unit] = parallelRounds.modify { rounds =>
+    val updatedIdRounds = rounds.get(id).map(r => r ++ roundIds).getOrElse(roundIds)
+
+    (rounds + (id -> updatedIdRounds), ())
+  }
+
+  def getRoundPeers(): F[Map[Id, RoundData]] = roundPeers.get
+
+  def addRoundPeer(id: Id, peerData: PeerData): F[Unit] =
+    roundPeers.modify(peers => (peers + (id -> RoundData(peerData.asRight[ProxyPeers])), ()))
+
+  def removePeer(id: Id): F[Unit] = roundPeers.modify(peers => (peers - id, ()))
+
+  def manageAbsentPeers(absentPeers: Set[Id]): F[Unit] =
+    for {
+      peers <- getRoundPeers()
+      // should we only remove peers that didn't finish or all of them
+      unsuccessfulPeers = peers.filter {
+        case (_, roundData) => !roundData.receivedProposal || roundData.consensusData.isEmpty
+      }.keySet
+      toRemove = unsuccessfulPeers.intersect(absentPeers)
+      _ <- leftDuringConsensus.modify(alreadyLeft => (alreadyLeft ++ toRemove, ()))
+      _ <- toRemove.toList.traverse(removePeer)
+    } yield ()
+
+  def manageProxiedPeers(peersThatNoLongerNeedProxying: Map[Id, PeerData]): F[Unit] =
```

Comment:
what is a proxy peer? shouldn't proposals be send using gossiping functionality?

Reply:
Didn't use gossiping in this implementation as it would be to complex - at least for me - and not sure if it would be correct to use it. Also deferring the failure to happen at other node, finding out about it and then handling it locally - not having it right by your hand when it's needed felt like it's a better to just broadcast.
The proxy peers are though used in a situation when some other peers sees the peers that is not connected to us. And we have it in order to get all the proposals - so we are sending and fetching the proposals through these proxyPeers. That's a corner case that I guess may at least theoretically happen.